src/ebmc/ranking_function.cpp

```diff
@@ -0,0 +1,216 @@
+
+/// \file
+/// Ranking Function Check
+
+#include "ranking_function.h"
+
+#include <trans-word-level/instantiate_word_level.h>
+#include <trans-word-level/property.h>
+#include <trans-word-level/trans_trace_word_level.h>
+#include <trans-word-level/unwind.h>
+
+#include <langapi/language.h>
+#include <langapi/mode.h>
+
+#include <solvers/flattening/boolbv.h>
+#include <solvers/sat/satcheck.h>
+
+#include <util/std_expr.h>
+
+#include "ebmc_base.h"
+
+/*******************************************************************\
+
+   Class: ranking_function_checkt
+
+ Purpose:
+
+\*******************************************************************/
+
+class ranking_function_checkt : public ebmc_baset
+{
+public:
+  ranking_function_checkt(
+    const cmdlinet &_cmdline,
+    ui_message_handlert &_ui_message_handler)
+    : ebmc_baset(_cmdline, _ui_message_handler), ns{symbol_table}
+  {
+  }
+
+  int operator()();
+
+protected:
+  namespacet ns;
+
+  exprt parse_ranking_function();
+  propertyt &find_property();
+};
+
+int do_ranking_function(
+  const cmdlinet &cmdline,
+  ui_message_handlert &ui_message_handler)
+{
+  return ranking_function_checkt(cmdline, ui_message_handler)();
+}
+
+exprt ranking_function_checkt::parse_ranking_function()
+{
+  auto language = get_language_from_mode(main_symbol->mode);
+  exprt expr;
+
+  language->set_message_handler(get_message_handler());
+
+  if(language->to_expr(
+       cmdline.get_value("ranking-function"),
+       id2string(main_symbol->module),
+       expr,
+       ns))
+  {
+    error() << "failed to parse ranking function" << eom;
+    return nil_exprt();
+  }
+
+  std::string expr_as_string;
+  language->from_expr(expr, expr_as_string, ns);
+  debug() << "Ranking function: " << expr_as_string << eom;
+  debug() << "Mode: " << main_symbol->mode << eom;
+
+  return expr;
+}
+
+ebmc_baset::propertyt &ranking_function_checkt::find_property()
+{
+  std::size_t count = 0;
+  for(auto &p : properties)
+  {
+    if(!p.is_disabled())
+      count++;
+  }
+
+  if(count == 0)
+  {
+    error() << "no property given" << eom;
+    throw 0;
+  }
+
+  if(count >= 2)
+  {
+    error() << "multiple properties are given -- please pick one" << eom;
+    throw 0;
+  }
+
+  for(auto &p : properties)
+  {
+    if(!p.is_disabled())
+      return p;
+  }
+
+  INVARIANT(false, "must not reach here");
+}
+
+int ranking_function_checkt::operator()()
+{
+  // get the transition system
+  int exit_code = get_model();
+
+  if(exit_code != -1)
+    return exit_code;
+
+  CHECK_RETURN(trans_expr.has_value());
+
+  // parse the ranking function
+  auto ranking_function = parse_ranking_function();
+
+  if(ranking_function.is_nil())
+    return 1;
+
+  // needs to have some numerical type
+  if(
+    ranking_function.type().id() != ID_signedbv &&
+    ranking_function.type().id() != ID_unsignedbv)
+  {
+    error() << "ranking function must have integral bitvector type" << eom;
+    return 1;
+  }
```

Comment:
Should this perhaps be part of `parse_ranking_function`?

Reply:
sure